File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int maxsofar = 0, maxcur = 0;
        for (int i = 1; i < prices.size(); i++) {
            maxcur = max(0, maxcur += prices[i] - prices[i - 1]);
            maxsofar = max(maxsofar, maxcur);
        }
        return maxsofar;
    }
};

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] p) {
        int maxSoFar = 0, maxCur = 0;
        for (int i = 1; i < p.length; i++) {
            maxCur = Math.max(0, maxCur += p[i] - p[i - 1]);
            maxSoFar = Math.max(maxSoFar, maxCur);
        }
        return maxSoFar;
    }
}